### ISSUE TITLE
JSON Parser - Integer values not working

### ISSUE THREAD

**Issue by syepes** [fluent/fluent-bit#1647]
## Bug Report

**Describe the bug**
Non string values (integers) are not correctly parsed.

**To Reproduce**
docker run --rm -it -v $PWD/fb.conf:/fluent-bit/etc/fluent-bit.conf -v $PWD/fb-parser.conf:/fluent-bit/etc/parsers.conf fluent/fluent-bit:1.3.2-debug

fb.conf
```
[SERVICE]
    Log_Level    trace
    Parsers_File parsers.conf

[INPUT]
    Name dummy
    Tag  dummy.data
    Dummy {"msg":"Non-Working","time":1570569413}

[INPUT]
    Name dummy
    Tag  dummy.data
    Dummy {"msg":"Working","time":"1570569413"}

[FILTER]
    Name parser
    Match dummy.*
    Key_Name data
    Parser dummy_test
    Reserve_Data On
    Preserve_Key On

[OUTPUT]
    Name stdout
    Match *
    Format json_lines
```

fb-parser.conf
```
[PARSER]
    Name        dummy_test
    Format      json
    Time_Key    time
    Time_Format %s
    Time_Keep   On
    Types       time:integer
```

**Expected behavior**
The parser "dummy_test" should be able to convert both epoch times (with and without double quotes)
 

**Screenshots**

```
Fluent Bit v1.3.2
Copyright (C) Treasure Data

[2019/10/11 20:08:05] [debug] [storage] [cio stream] new stream registered: dummy.0
[2019/10/11 20:08:05] [debug] [storage] [cio stream] new stream registered: dummy.1
[2019/10/11 20:08:05] [ info] [storage] initializing...
[2019/10/11 20:08:05] [ info] [storage] in-memory
[2019/10/11 20:08:05] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2019/10/11 20:08:05] [ info] [engine] started (pid=1)
[2019/10/11 20:08:05] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2019/10/11 20:08:05] [debug] [router] match rule dummy.0:stdout.0
[2019/10/11 20:08:05] [debug] [router] match rule dummy.1:stdout.0
[2019/10/11 20:08:05] [ info] [sp] stream processor started
[2019/10/11 20:08:09] [debug] [task] created task=0x7f20d60471e0 id=0 OK
{"date":1570824486.000213,"msg":"Non-Working","time":1570569413}
{"date":1570824487.000182,"msg":"Non-Working","time":1570569413}
{"date":1570824488.000178,"msg":"Non-Working","time":1570569413}
{"date":1570824489.000201,"msg":"Non-Working","time":1570569413}
[2019/10/11 20:08:09] [debug] [task] created task=0x7f20d6047280 id=1 OK
{"date":1570824486.000313,"msg":"Working","time":"1570569413"}
{"date":1570824487.000233,"msg":"Working","time":"1570569413"}
{"date":1570824488.000233,"msg":"Working","time":"1570569413"}
{"date":1570824489.000254,"msg":"Working","time":"1570569413"}
[2019/10/11 20:08:10] [debug] [task] destroy task=0x7f20d60471e0 (task_id=0)
[2019/10/11 20:08:10] [debug] [task] destroy task=0x7f20d6047280 (task_id=1)

```

**Comment by nokute78** [comment]
First, in_dummy generates MessagePack , not JSON.
JSON parser is to parse the JSON string in MessagePack.

e.g. 
```
Dummy {"json_str":"{"msg":"Working","time":1570569413}"}
```
In this case, the parser is to parse the field of 'json_str'.

a.conf:
```
[SERVICE]
    Parsers_File parsers.conf

[INPUT]
    Name dummy
    Tag  dummy.data
    Dummy {"json_str":"{\"msg\":\"Non-Working\",\"time\":1570569413}"}

[FILTER]
    Name parser
    Match dummy.*
    Key_Name json_str
    Parser dummy_test
    Reserve_data on
    Preserve_key on

[OUTPUT]
    Name stdout
    Match *
```

parsers.conf:
```
[PARSER]
    Name        dummy_test
    Format      json
```

The result is
```
$ ../bin/fluent-bit -c a.conf 
Fluent Bit v1.4.0
Copyright (C) Treasure Data

[2019/10/12 08:51:50] [ info] [storage] initializing...
[2019/10/12 08:51:50] [ info] [storage] in-memory
[2019/10/12 08:51:50] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2019/10/12 08:51:50] [ info] [engine] started (pid=32479)
[2019/10/12 08:51:50] [ info] [sp] stream processor started
[0] dummy.data: [1570837911.000278008, {"msg"=>"Non-Working", "time"=>1570569413, "json_str"=>"{"msg":"Non-Working","time":1570569413}"}]
[1] dummy.data: [1570837912.001061238, {"msg"=>"Non-Working", "time"=>1570569413, "json_str"=>"{"msg":"Non-Working","time":1570569413}"}]
[2] dummy.data: [1570837913.001026635, {"msg"=>"Non-Working", "time"=>1570569413, "json_str"=>"{"msg":"Non-Working","time":1570569413}"}]
[3] dummy.data: [1570837914.000981615, {"msg"=>"Non-Working", "time"=>1570569413, "json_str"=>"{"msg":"Non-Working","time":1570569413}"}]
```

Second, JSON parser doesn't support `Types` option. 
See also
https://github.com/fluent/fluent-bit/issues/1563#issuecomment-531749098

**Comment by syepes** [comment]
@nokute78 Thanks for the pointers on the MessagePack stuff, I was not aware of that.
I have adapted by example and the original issue is still present:

fb.conf
```
[SERVICE]
    Log_Level    trace
    Parsers_File parsers.conf

[INPUT]
    Name  dummy
    Tag   dummy.data
    Dummy {"json_str":"{\"msg\":\"Non-Working\",\"time\":1570569410}"}

[INPUT]
    Name  dummy
    Tag   dummy.data
    Dummy {"json_str":"{\"msg\":\"Working\",\"time\":\"1570569413\"}"}


[FILTER]
    Name parser
    Match dummy.*
    Key_Name json_str
    Parser dummy_test
    Reserve_Data On
    Preserve_Key On

[OUTPUT]
    Name stdout
    Match *
    Format json_lines
```

fb-parser.conf
```
[PARSER]
    Name        dummy_test
    Format      json
    Time_Key    time
    Time_Format %s        # If we add this it fails (epoch)
```


The result is
```
ocker run --rm -it -v $PWD/fb.conf:/fluent-bit/etc/fluent-bit.conf -v $PWD/fb-parser.conf:/fluent-bit/etc/parsers.conf fluent/fluent-bit:1.3.2-debug
Fluent Bit v1.3.2
Copyright (C) Treasure Data

[2019/10/12 05:37:43] [debug] [storage] [cio stream] new stream registered: dummy.0
[2019/10/12 05:37:43] [debug] [storage] [cio stream] new stream registered: dummy.1
[2019/10/12 05:37:43] [ info] [storage] initializing...
[2019/10/12 05:37:43] [ info] [storage] in-memory
[2019/10/12 05:37:43] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2019/10/12 05:37:43] [ info] [engine] started (pid=1)
[2019/10/12 05:37:43] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2019/10/12 05:37:43] [debug] [router] match rule dummy.0:stdout.0
[2019/10/12 05:37:43] [debug] [router] match rule dummy.1:stdout.0
[2019/10/12 05:37:43] [ info] [sp] stream processor started
[2019/10/12 05:37:43] [error] [parser] time string length is too long
[2019/10/12 05:37:43] [ warn] [parser:dummy_test] Invalid time format %s    # test for 'time�]���'.
[2019/10/12 05:37:44] [ warn] [parser:dummy_test] Invalid time format %s    # test for '1570569413'.
[2019/10/12 05:37:44] [error] [parser] time string length is too long
[2019/10/12 05:37:44] [ warn] [parser:dummy_test] Invalid time format %s    # test for 'time�]��¦errors'.
[2019/10/12 05:37:44] [ warn] [parser:dummy_test] Invalid time format %s    # test for '1570569413'.

```


Essentially what I want to achieve is to extract the epoch time from the json (without the double quotes) and use that as the timestamp of the event.

Is this currently possible?

**Comment by nokute78** [comment]
@syepes I remember the issue #662 . But we can convert now with filter_lua.

time.lua:
```lua
function time2str(tag, timestamp, record)
         return 1, record["time"], record
end
```
`1` means lua script changed timestamp/record.
`record["time"]` is the value of `time` which is extracted with JSON parser.
In this case, it is `1570569413`.

a.conf
```
[SERVICE]
    Parsers_File parsers.conf

[INPUT]
    Name dummy
    Tag  dummy.data
    Dummy {"json_str":"{\"msg\":\"Non-Working\",\"time\":1570569413}"}

[FILTER]
    Name parser
    Match dummy.*
    Key_Name json_str
    Parser dummy_test
    Reserve_data on
    Preserve_key on

[FILTER]
    Name lua
    Match dummy.*
    script time.lua
    call   time2str

[OUTPUT]
    Name stdout
    Match *
```

parsers.conf
```
[PARSER]
    Name        dummy_test
    Format      json
```

result is
```
$ ../bin/fluent-bit -c a.conf 
Fluent Bit v1.4.0
Copyright (C) Treasure Data

[2019/10/13 06:36:31] [ info] [storage] initializing...
[2019/10/13 06:36:31] [ info] [storage] in-memory
[2019/10/13 06:36:31] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2019/10/13 06:36:31] [ info] [engine] started (pid=35118)
[2019/10/13 06:36:31] [ info] [sp] stream processor started
[0] dummy.data: [1570569413.000000000, {"msg"=>"Non-Working", "json_str"=>"{"msg":"Non-Working","time":1570569413}", "time"=>1570569413.000000}]
[1] dummy.data: [1570569413.000000000, {"msg"=>"Non-Working", "json_str"=>"{"msg":"Non-Working","time":1570569413}", "time"=>1570569413.000000}]
[2] dummy.data: [1570569413.000000000, {"msg"=>"Non-Working", "json_str"=>"{"msg":"Non-Working","time":1570569413}", "time"=>1570569413.000000}]
[3] dummy.data: [1570569413.000000000, {"msg"=>"Non-Working", "json_str"=>"{"msg":"Non-Working","time":1570569413}", "time"=>1570569413.000000}]
^C[engine] caught signal (SIGINT)
[2019/10/13 06:36:36] [ info] [input] pausing dummy.0
```

**Comment by syepes** [comment]
@nokute78 
Great this was resolved the issue, but at the same time has added another :-)
All other fields that contain integers like (value) in the below example, are now converted to floats.

Is there a simple way to prevent this? of just to affect the time field.

a.conf
```
[INPUT]
    Name  dummy
    Tag   dummy.data
    Dummy {"json_str":"{\"msg\":\"Non-Working\",\"time\":1570569410, \"value\":1}"}

[INPUT]
    Name  dummy
    Tag   dummy.data
    Dummy {"json_str":"{\"msg\":\"Working\",\"time\":\"1570569413\", \"value\":2}"}
```

result is
```
Fluent Bit v1.3.2
Copyright (C) Treasure Data
{"date":1570569410.000000,"msg":"Non-Working","time":1570569410.000000,"value":1.000000}
{"date":1570569410.000000,"msg":"Non-Working","time":1570569410.000000,"value":1.000000}
{"date":1570569413.000000,"msg":"Working","time":"1570569413","value":2.000000}
{"date":1570569413.000000,"msg":"Working","time":"1570569413","value":2.000000}
```

**Comment by nokute78** [comment]
@syepes 
> a field that can have both types of values

Would you share an example / usecase ?

It may be better to treat as float.
e.g. Fixed data type is convenient for database.

**Comment by edsiper** [comment]
About the problem that numbers were converted to double, this is addressed in c8f55c8b7850ca07c51002cfa2040642111eee18 . 

The solution will be part of Fluent Bit v1.5.7
